### ISSUE TITLE
set_weights() in models.py is not working.

### ISSUE THREAD

**Issue by verystrongjoe** [keras-team/keras#7906]
I am running [this source code](https://github.com/jjakimoto/DQN). But It was build by Keras 1.x. So I am running Keras 2.0.6 so I changed some source code to make it running on new version of Keras. But I kept failing copying the weights of critic model to the weight of target critic model. Those two models were built by same function below. I can not understand why this error get occurs.

** copy weights logic **
```{python}
        actor_model_weights = self.critic.get_weights()
        actor_target_weights = self.critic_target.get_weights()

        for i in range(len(actor_target_weights)):
            actor_target_weights[i] = actor_model_weights[i]
        self.critic_target.set_weights(actor_target_weights)
```

**error**  
```
  File "C:\Program Files\Anaconda3\envs\tensorflow\lib\site-packages\keras\models.py", line 698, in set_weights
    layer.set_weights(weights[:nb_param])
  File "C:\Program Files\Anaconda3\envs\tensorflow\lib\site-packages\keras\engine\topology.py", line 1186, in set_weights
    if pv.shape != w.shape:
AttributeError: 'list' object has no attribute 'shape'

Process finished with exit code 1
```

**Function to build critic model** 
```
 def build_critic(self):
        """Build critic network
        recieve transformed tensor: raw_data, smooted_data, and downsampled_data
        """
        nf = self.n_feature
        # layer1
        # smoothed input
        sm_model = [Sequential() for _ in range(self.n_smooth)]
        for m in sm_model:
            m.add(Lambda(lambda x: x,  input_shape=(self.history_length, self.n_stock, 1)))
            m.add(Convolution2D(nb_filter=nf, nb_row=self.k_w, nb_col=1, border_mode='same'))
            #m.add(BatchNormalization(mode=2, axis=-1))
            m.add(BatchNormalization(axis=-1))
            m.add(PReLU())
        # down sampled input
        dw_model = [Sequential() for _ in range(self.n_down)]
        for m in dw_model:
            m.add(Lambda(lambda x: x,  input_shape=(self.history_length, self.n_stock, 1)))
            m.add(Convolution2D(nb_filter=nf, nb_row=self.k_w, nb_col=1, border_mode='same'))
            #m.add(BatchNormalization(mode=2, axis=-1))
            m.add(BatchNormalization(axis=-1))
            m.add(PReLU())

        # raw input
        state = Sequential()
        nf = self.n_feature
        state.add(Lambda(lambda x: x,  input_shape=(self.history_length, self.n_stock, 1)))
        state.add(Convolution2D(nb_filter=nf, nb_row=self.k_w, nb_col=1, border_mode='same'))
        # state.add(BatchNormalization(mode=2, axis=-1))
        state.add(BatchNormalization(axis=-1))
        state.add(PReLU())

        merged = Merge([state, ] + sm_model + dw_model, 'concat', -1)
        #https: // stackoverflow.com / questions / 43196636 / how - to - concatenate - two - layers - in -keras

        # layer2
        nf = nf * 2
        model = Sequential()
        model.add(merged)
        model.add(Convolution2D(nb_filter=nf, nb_row=self.k_w, nb_col=1, border_mode='same'))
        #model.add(BatchNormalization(mode=2, axis=-1))
        model.add(BatchNormalization(axis=-1))
        model.add(PReLU())
        model.add(Flatten())
        # layer3
        model.add(Dense(self.n_hidden))
        #model.add(BatchNormalization(mode=1, axis=-1))
        model.add(BatchNormalization(axis=-1))
        model.add(PReLU())
        # layer4
        model.add(Dense(int(np.sqrt(self.n_hidden))))
        model.add(PReLU())
        # output
        model.add(Dense(2 * self.n_stock))
        model.add(Reshape((self.n_stock, 2)))
        return model
```

Thanks for reading!

**Comment by drscotthawley** [comment]
Don't have an answer, just confirming a similar problem:  Trying to copy weights from 'model' to an identically-defined  'new_model' fails...

`new_model.set_weights( model.get_weights() ) 
`

> Traceback (most recent call last):
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/framework/common_shapes.py", line 671, in _call_cpp_shape_fn_impl
>     input_tensors_as_shapes, status)
>   File "/opt/anaconda/envs/py35/lib/python3.5/contextlib.py", line 66, in __exit__
>     next(self.gen)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/framework/errors_impl.py", line 466, in raise_exception_on_not_ok_status
>     pywrap_tensorflow.TF_GetCode(status))
> tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimension 1 in both shapes must be equal, but are 224 and 3 for 'Assign_1792' (op: 'Assign') with input shapes: [38400,224], [38400,3].
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>     new_model.set_weights( model.get_weights() )
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/keras/engine/topology.py", line 2004, in set_weights
>     K.batch_set_value(tuples)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/keras/backend/tensorflow_backend.py", line 2188, in batch_set_value
>     assign_op = x.assign(assign_placeholder)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/ops/variables.py", line 512, in assign
>     return state_ops.assign(self._variable, value, use_locking=use_locking)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/ops/state_ops.py", line 270, in assign
>     validate_shape=validate_shape)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/ops/gen_state_ops.py", line 47, in assign
>     use_locking=use_locking, name=name)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/framework/op_def_library.py", line 768, in apply_op
>     op_def=op_def)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 2338, in create_op
>     set_shapes_for_outputs(ret)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 1719, in set_shapes_for_outputs
>     shapes = shape_func(op)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/framework/ops.py", line 1669, in call_with_requiring
>     return call_cpp_shape_fn(op, require_shape_fn=True)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/framework/common_shapes.py", line 610, in call_cpp_shape_fn
>     debug_python_shape_fn, require_shape_fn)
>   File "/opt/anaconda/envs/py35/lib/python3.5/site-packages/tensorflow/python/framework/common_shapes.py", line 676, in _call_cpp_shape_fn_impl
>     raise ValueError(err.message)
> ValueError: Dimension 1 in both shapes must be equal, but are 224 and 3 for 'Assign_1792' (op: 'Assign') with input shapes: [38400,224], [38400,3].
> 
>

**Comment by alexantoinefortin** [comment]
+1 I have the same error message now that I just added a BatchNormalization layer to my model.
Before adding this BatchNormalization layer, it used to run fine.

My model is set up similar to yours (critic network + critic target network)

**Comment by alexantoinefortin** [comment]
So after investigation and for an unknown reason, the weights of my BatchNormalization layer were returned as a list of lists (eg: `[[1,2,3],[1,2,3],[1,2,3]]`) whereas for my other layers, I was getting a list of numpy arrays (eg:`[array([1,2,3]),array([1,2,3]),array([1,2,3])]`).

I ended up fixing this problem with a list comprehension.
If I were to apply my fix to your code, it would look like this:
  ```python
for i in range(len(actor_target_weights)):
    actor_target_weights[i] = actor_model_weights[i]
    if ((len(actor_target_weights)>0) and (type(actor_target_weights[0])==type([1,2])))   
        actor_target_weights = [np.array(x) for x in actor_target_weights]
    self.critic_target.set_weights(actor_target_weights)
```

Ugly, but it seems to work.

Note, in my code, I was using np.multiply to perform an element-wise multiplication of 2 list of np.arrays. Only after the np.multiply step did my BatchNormalization layer's weights became a list of lists. Before the np.multiply, I had a list of np.array as my BatchNormalization layer.

Out of 5-6 layers, only the BatchNormalization layer was changed from a `list of np.array` to a `list of lists` after my np.multiply step.

I hope it helps!
